klib/cloudwatch.c

```diff
@@ -256,39 +583,243 @@ define_closure_function(0, 1, input_buffer_handler, cw_metrics_conn_handler,
+        spin_lock(&cw.lock);
+        cw_log_send_async();
+        spin_unlock(&cw.lock);
     }
 }
 
-define_closure_function(0, 2, void, cw_metrics_timer_handler,
+static boolean cw_log_token_get(buffer resp, const char *label)
+{
+    int tok_start = buffer_strstr(resp, label);
+    if (tok_start >= 0) {
+        buffer_consume(resp, tok_start + runtime_strlen(label));
+        tok_start = buffer_strchr(resp, '"');
+        if (tok_start >= 0) {
+            buffer_consume(resp, tok_start + 1);
+            int tok_end = buffer_strchr(resp, '"');
+            if (tok_end >= 0) {
+                buffer_clear(cw.log_seq_token);
+                return buffer_write(cw.log_seq_token, buffer_ref(resp, 0), tok_end);
+            }
+        }
+    }
+    return false;
+}
+
+static void cw_log_pending_delete(void)
+{
+    spin_lock(&cw.lock);
+    int log_entries = cw.log_pending;
+    for (int i = 0; i < log_entries; i++)
+        deallocate(cw.h, vector_get(cw.log_entries, i), sizeof(struct cw_log_entry));
+    vector_delete_range(cw.log_entries, 0, log_entries);
+    spin_unlock(&cw.lock);
+}
+
+define_closure_function(0, 1, void, cw_log_vh,
+                        value, v)
+{
+    buffer content = get_string(v, sym(content));
+    if (content) {
+        if (cw_log_token_get(content, CW_LOG_NEXT_SEQ_TOK)) {
+            cw_log_pending_delete();
+        } else if (cw_log_token_get(content, CW_LOG_EXP_SEQ_TOK)) {
+            /* resend log entries at the next timer interval */
+        } else if (buffer_strstr(content, AWS_ERR_TOKEN_EXPIRED) >= 0) {
+            aws_cred_get(cw.h, init_closure(&cw.aws_cred_handler, cw_aws_cred_handler,
+                         (status_handler)&cw.aws_setup_complete));
+        } else {
+            msg_err("unexpected response %v\n", v);
+            cw_log_pending_delete();
+        }
+    }
+    cw.log_resp_recved = true;
+}
+
+define_closure_function(0, 1, boolean, cw_log_in_handler,
+                        buffer, data)
+{
+    if (data) {
+        status s = apply(cw.log_resp_parser, data);
+        if (is_ok(s)) {
+            if (cw.log_resp_recved) {
+                apply(cw.log_out, 0);
+                return true;
+            }
+        } else {
+            msg_err("failed to parse response: %v\n", s);
+            timm_dealloc(s);
+            apply(cw.log_out, 0);
+            return true;
+        }
+    } else {    /* connection closed */
+        spin_lock(&cw.lock);
+        cw.log_pending = 0;
+        cw.log_out = 0;
+        cw_log_send_async();
+        spin_unlock(&cw.lock);
+    }
+    return false;
+}
+
+static void cw_log_post(void)
+{
+    tuple req = allocate_tuple();
+    if (req == INVALID_ADDRESS)
+        goto error;
+    set(req, sym(url), alloca_wrap_cstring("/"));
+    set(req, sym(host), alloca_wrap_cstring(cw.log_servername));
+    set(req, sym(x-amz-target), alloca_wrap_cstring("Logs_20140328.PutLogEvents"));
```

Comment:
The "x-amz-target" string itself is not specific to the logs API, it's used across many AWS APIs, not limited to CloudWatch. I don't think it will ever need to be overridden. I changed the cw_aws_req_send() function to take an additional `const char *target` argument and do `set(req, sym(x-amz-target), alloca_wrap_cstring(target))` at the beginning, so this code is not duplicated for each function call.